### PR TITLE
[SPARK-8246] [SQL] Implement get_json_object

### PR DESCRIPTION
**Pull request by NathanHowell**
First stab at implementing get_json_object... had some time to kill last night and hammered it out. I haven't looked at where this integrates into the parser (`FunctionRegistry` maybe?) but it works in manually built up DataFrames.

Tests were copied from Hive, but do not seem to be super comprehensive. I've generally replicated Hive's unusual behavior rather than following a JSONPath reference, except for one case (as noted in the comments). I don't know if there is a way of fully replicating Hive's behavior without a slower TreeNode implementation, so I've erred on the side of performance instead.

### CODE REVIEW

**Comment by rxin** [comment]
Yup you can add it to FunctionRegistry and it will be available in the parser automatically after that.

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by NathanHowell** [comment]
Bah, scalastyle fix didn't get rebuilt by sbt. A one char fix in the failed test, and the `FunctionRegistry` modification are in now. I'll be out for a few days, should be back online by Tuesday if anything else needs to be massaged.

**Comment by rxin** [comment]
Jenkins, ok to test.

**Comment by marmbrus** [comment]
ok to test

**Comment by NathanHowell** [comment]
Rebased against master.

I figured out an easy way to regain 100% compatibility with Hive's tests... I'll take a stab at a fix tonight or tomorrow.

**Comment by NathanHowell** [comment]
I haven't run the full Hive compatibility test suite yet, the test build may fail.

**Comment by NathanHowell** [comment]
Humph, been having trouble getting the Hive compatibility tests running under a debugger. I should have time midweek to dig in further.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonFunctionsSuite.scala [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.SparkFunSuite
+
+class JsonFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper {
```
Comment (davies) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonFunctionsSuite.scala]: This should be called JsonExpressionsSuite, JsonFunctionsSuite is usually used for end-to-end tests for DataFrame API and SQL (using df.selectExpr()), would be nice to have one for this function.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetJsonObject.scala [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+  private val jsonFactory = new JsonFactory()
+
+  // Enabled for Hive compatibility
+  jsonFactory.enable(JsonParser.Feature.ALLOW_UNQUOTED_CONTROL_CHARS)
+}
+
+case class GetJsonObject(
+    jsonExpression: Expression,
+    pathExpression: Expression)
+  extends Expression
```
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/GetJsonObject.scala]: it could extends BinaryExpression

**Comment by davies** [comment]
@NathanHowell Thanks for the hard work, it looks good to me in general, it would be cool to have this in 1.5 release, would you mind me to take over this one to fix the Hive test and some minor code style?

**Comment by NathanHowell** [comment]
@davies go ahead and take it over, I'd like to see it in 1.5 as well... I killed another hour trying to get the Hive tests running locally last week without any real progress :disappointed:

**Comment by davies** [comment]
@NathanHowell This is merged, could you close it? thanks

**Comment by NathanHowell** [comment]
Done, thanks for the help!

**Pull request closed without merging**